### PR TITLE
Add `beforeRow` and `afterRow`

### PR DESCRIPTION
**Pull request by mdmunir**
| Q | A |
| --- | --- |
| Is bugfix? | no |
| New feature? | yes |
| Breaks BC? | no |
| Tests pass? |  |
| Fixed issues |  |
# Usage

``` php
ListView::widget([
    'dataProvider' => $dataProvider,
    'itemOptions' => ['class' => 'col-md-4'],
    'beforeRow' => function($model, $key, $index) {
        if ($index % 3 === 0) {
            return '<div class="row">';
        }
    },
   'afterRow' => function($model, $key, $index) use($dataProvider) {
        if ($index % 3 === 2 || $index === $dataProvider->count - 1) {
            return '</div>';
        }
    },
]);
```

### CODE REVIEW

**Comment by samdark** [comment]
I have no objections merging it. @yiisoft/core-developers opinions?

File: framework/widgets/ListView.php [hunk 1]
```diff
@@ -62,6 +62,28 @@ class ListView extends BaseListView
...
+     * ```
+     *
+     * - `$model`: the current data model being rendered
+     * - `$key`: the key value associated with the current data model
+     * - `$index`: the zero-based index of the data model in the model array returned by [[dataProvider]]
+     * - `$widget`: the ListView object
+     *
+     * The return result of the function will be rendered directly.
+     */
+    public $beforeRow;
```
Comment (klimov-paul) [inline framework/widgets/ListView.php]: Should be `beforeItem`.

File: framework/widgets/ListView.php [hunk 2]
```diff
@@ -62,6 +62,28 @@ class ListView extends BaseListView
...
+    /**
+     * @var Closure an anonymous function that is called once AFTER rendering each data model.
+     * It should have the similar signature as [[beforeRow]]. The return result of the function
+     * will be rendered directly.
+     */
+    public $afterRow;
```
Comment (klimov-paul) [inline framework/widgets/ListView.php]: Should be `afterItem`

**Comment by mdmunir** [comment]
https://github.com/yiisoft/yii2/blob/master/framework/grid/GridView.php#L113-L119

File: framework/widgets/ListView.php (hunk 1 above)
Comment (klimov-paul) [inline framework/widgets/ListView.php]: Missing `since` tag

File: framework/widgets/ListView.php (hunk 2 above)
Comment (klimov-paul) [inline framework/widgets/ListView.php]: Missing `since` tag

File: framework/widgets/ListView.php [hunk 3]
```diff
@@ -74,7 +96,22 @@ public function renderItems()
         $keys = $this->dataProvider->getKeys();
         $rows = [];
         foreach (array_values($models) as $index => $model) {
-            $rows[] = $this->renderItem($model, $keys[$index], $index);
+            $key = $keys[$index];
+            if ($this->beforeRow !== null) {
+                $row = call_user_func($this->beforeRow, $model, $key, $index, $this);
+                if (!empty($row)) {
```
Comment (klimov-paul) [inline framework/widgets/ListView.php]: Should be check for `null` or `false`.
Empty string added to the items set may be a desiarable behavior.

**Review by klimov-paul — CHANGES_REQUESTED** [review]

**Comment by cebe** [comment]
Looks good to me, need a CHANELOG entry.

**Review by samdark — APPROVED** [review]

**Review by klimov-paul — APPROVED** [review]

**Comment by samdark** [comment]
@yiisoft/core-developers 2.0.10 or 2.0.11?

**Comment by klimov-paul** [comment]
I am for `2.0.11`.
`ListView` is not covered by unit tests. It is better to move this fix to the beginning of bew milestone.

**Comment by dynasource** [comment]
> ListView is not covered by unit tests
Im willing to fix that

**Comment by SilverFire** [comment]
Merged with some adjustments: https://github.com/yiisoft/yii2/commit/0148325426ea495cbfcc7170ed8c45a0bccfbb1e

**Comment by SilverFire** [comment]
Thank you!

**Pull request closed without merging**